// nerdcord/dist/src/core/client/Client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const WebSocketManager_1 = __importDefault(require("../ws/WebSocketManager"));
const events_1 = require("events");
const node_fetch_1 = __importDefault(require("node-fetch"));
class Client extends events_1.EventEmitter {
    constructor() {
        super(...arguments);
        // @ts-ignore
        this.socket = new WebSocketManager_1.default(this);
    }
    async login(token) {
        if (!token)
            throw new Error("please provide a token");
        this.socket.login(token);
    }
    async createMessage(content, channelID, token) {
        const data = {
            "content": content,
            "tts": false
        };
        const headers = { 'Content-Type': 'application/json', 'Authorization': `Bot ${token}` };
        const response = await node_fetch_1.default(`https://discord.com/api/v6/channels/${channelID}/messages`, {
            method: "POST",
            headers,
            body: JSON.stringify(data)
        });
        const res = await response.json();
        console.log(res);
        return response;
    }
    async createEmbed(description, channelID, token, title) {
        const data = {
            "tts": false,
            "embed": {
                "title": title,
                "description": description
            }
        };
        const headers = { 'Content-Type': 'application/json', 'Authorization': `Bot ${token}` };
        const response = await node_fetch_1.default(`https://discord.com/api/v6/channels/${channelID}/messages`, {
            method: "POST",
            headers,
            body: JSON.stringify(data)
        });
        return response;
    }
    async addGuildBan() {
    }
}
exports.default = Client;
