// nerdcord/dist/src/core/ws/WebSocketManager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ws_1 = __importDefault(require("ws"));
const Payloads_1 = require("../../constants/Payloads");
const Constants_1 = require("../../constants/Constants");
class WebSocketManager {
    constructor(client) {
        this.client = client;
    }
    async login(token) {
        const properties = { $os: 'linux', $browser: 'nerdcord', $device: 'nerdcord' };
        const identify = {
            op: 2,
            d: { token, properties }
        };
        const GATEWAY = "wss://gateway.discord.gg/?v=6&encoding=json";
        this.socket = await new ws_1.default(GATEWAY);
        this.socket.on("open", () => {
            //console.log("connected")
        });
        /*this.socket.onopen = () => {
            setInterval(() => {
                this.socket.send(JSON.stringify(identify))
            }, 41250)
        }*/
        this.socket.on("message", async (message) => {
            const payload = JSON.parse(message.toString());
            const { t: event, s, op, d } = payload;
            const { heartbeat_interval } = d;
            // console.log(payload)
            switch (payload.op) {
                case Constants_1.OPCODE.ZERO:
                    break;
                case Constants_1.OPCODE.TEN:
                    this.interval = this.heartbeat(heartbeat_interval);
                    await this.identify(token);
                    break;
                case Constants_1.OPCODE.ELEVEN:
                    break;
            }
            if (event) {
                try {
                    const { default: module } = await Promise.resolve().then(() => __importStar(require(`../handlers/${event}`)));
                    module(this.client, payload);
                }
                catch (e) {
                    console.log(e);
                }
            }
        });
    }
    heartbeat(ms) {
        return setInterval(() => {
            this.socket.send(JSON.stringify(Payloads_1.Heartbeat));
        }, ms);
    }
    async identify(token) {
        Payloads_1.Identify.d.token = token;
        return this.socket.send(JSON.stringify(Payloads_1.Identify));
    }
}
exports.default = WebSocketManager;
